Make base images multiarch

martin-g opened the pull request:
Add support for Linux ARM64 to `base-glibc-busybox-bash` and `base-glibc-debian-bash` images.

Update the Github runners to 20.04 because the 18.04 ones never start (they stay in `Queued` state for days). See https://github.blog/changelog/2022-08-09-github-actions-the-ubuntu-18-04-actions-runner-image-is-being-deprecated-and-will-be-removed-by-12-1-22/

Fix warnings:
- update actions/checkout to v4, so that it does not use the deprecated Nodejs v12
- use `>> $GITHUB_OUTPUT` instead of the deprecated `set-output`

Multi-arch images built by my fork could be seen here:
- https://github.com/martin-g/bioconda-containers/pkgs/container/base-glibc-busybox-bash
- https://github.com/martin-g/bioconda-containers/pkgs/container/base-glibc-debian-bash

martin-g committed:
Add support for multiplatform to the Docker images

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

Try with redhat-actions/buildah-build and redhat-actions/push-to-registry

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Make busybox image multiarch

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented on .github/workflows/base-glibc-debian-bash.yaml:
```diff
@@ -93,8 +102,8 @@ jobs:
         respone="$(
-          curl -sL \
-            'https://quay.io/api/v1/repository/bioconda/${{ steps.buildah-build.outputs.image }}/image'
+          curl -sL -H "Authorization: Bearer ${{ secrets.QUAY_BIOCONDA_TOKEN }}" \
```
This is not tested because the secrets are not exported to forks/PRs.
Without the `Authorization` header Quay returns 403 (Forbidden)

martin-g committed:
Install wget for create-env Dockerfile

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Use ubuntu-20.04 for bioconda-recipes-issue-responder and bot workflows

ubuntu-18.04 is deprecated - https://github.blog/changelog/2022-08-09-github-actions-the-ubuntu-18-04-actions-runner-image-is-being-deprecated-and-will-be-removed-by-12-1-22/

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

Yikun commented on images/create-env/Dockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG arch=x86_64
 # Use the exact conda, mamba versions as used in bioconda-recipes' builds.
 ARG bioconda_utils_version='1.1.3'
 FROM quay.io/bioconda/bioconda-utils-build-env-cos7:${bioconda_utils_version} as bioconda-build-env
```
It seems the image should be `quay.io/bioconda/bioconda-utils-build-env-cos7` for x86 and `quay.io/bioconda/bioconda-utils-build-env-cos7-aarch64` for aarch64?

[1] https://github.com/bioconda/bioconda-utils/pull/866/files#diff-8b50c1c8ac07bc956388add66e256e8336ae22fd2cac8d83f770e1dc63674c00R20-R25

martin-g replied:
Yes, this has to be updated once https://github.com/bioconda/bioconda-utils/pull/866 is merged!

Yikun commented on images/create-env/Dockerfile:
```diff
@@ -13,7 +14,7 @@ FROM quay.io/bioconda/base-glibc-debian-bash as build
 COPY --from=bioconda-build-env /tmp/requirements.txt ./
 COPY install-conda print-env-activate create-env ./
-ADD https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh ./miniconda.sh
+ADD https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-${arch}.sh ./miniconda.sh
```
curious is it something special so that we have to use `build-args` but not wget according build arch?

martin-g replied:
Again, this is outdated. Please see the latest version of the PR.

martin-g replied:
But now I notice that I need to add multi-arch build to the GHA workflow!

martin-g replied:
Improved with https://github.com/bioconda/bioconda-containers/pull/55/commits/1b4b7e2dbac4313f006e6dd1220848ff6b3b3d0b

martin-g commented on .github/workflows/base-glibc-debian-bash.yaml:
```diff
@@ -14,19 +14,24 @@ on:
+      IMAGE_VERSION: '3.0.0'
       IMAGE_NAME: base-glibc-debian-bash
-      DEBIAN_VERSION: '10.9'
+      DEBIAN_VERSION: '10.12'
```
Outdated. Please see the latest version.

martin-g commented on .github/workflows/base-glibc-debian-bash.yaml:
```diff
@@ -1,4 +1,4 @@
-name: 'Build & Push: base-glibc-debian-bash'
+name: 'Build & Push: base-glibc-debian-bash 2'
```
Outdated. Please see the latest version.

martin-g committed:
Add multi-arch build to create-env workflow

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Pass an Authorization request header when making REST API calls to Quay.io

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

Yikun commented on .github/workflows/base-glibc-busybox-bash.yaml:
```diff
@@ -14,20 +14,25 @@ on:
       # The base image is not intended to change often and should be used with
       # version tags or checksum IDs, but not via "latest".
-      IMAGE_VERSION: '2.1.0'
+      IMAGE_VERSION: '3.0.0'
```
Is it a normal upgrade or for special fixing?

martin-g replied:
I bumped the major version because now the image would be multi-arch and because of the update of Ubuntu version and the GLIBC coming with it.
But since there are no changes for the x86_64 users maybe we should bump the minor/patch version ?! I am OK to update the PR with the preferred change by the maintainers!

bgruening replied:
The only reason I can think of to keep the older version is to use an older libc for testing to make sure our packages for on older HPC. But this should be ensure with the build container ... so following this and trusting the build container I think we could bump the runtime container.

@bioconda/build-system any thoughts?

martin-g replied:
I just bumped the version of `create-env` image too to `3.0.0`, so it is in sync with the other images in this PR.
As I said above I'd be happy to change the version to whatever the Bioconda team thinks it should be!

martin-g committed:
Bump create-env image version to 3.0.0

To be in sync with the other image versions in this PR

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
We are ready with the next PR for bioconda-containers - https://github.com/martin-g/bioconda-containers/pull/2/files
It depends on this PR to be merged and the new multiarch container images to be available at Quay.io. And also depends on the new `bioconda-utils-build-env-cos7-aarch64` which will be provided by https://github.com/bioconda/bioconda-utils/pull/866

The new/next PR will provide images `create-env-x86_64` and `create-env-aarch64` (https://github.com/martin-g?tab=packages&tab=packages&q=create-)

martin-g committed:
create-env container image is not ready yet for Linux ARM64

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

Yikun commented on .github/workflows/create-env.yaml:
```diff
@@ -16,11 +16,16 @@ jobs:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
+
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v2
+      with:
+        platforms: arm64
```
nit: seems it should be moved into arm64 PR?

martin-g replied:
Right! Removed it!

Yikun commented on .github/workflows/base-glibc-debian-bash.yaml:
```diff
@@ -44,10 +49,10 @@ jobs:
-        container="$( buildah from "${image_id}" )"
-        buildah config --label=glibc="${glibc}" "${container}"
-        buildah config --label=debian="${debian}" "${container}"
+        labels="
```
buildah config --label=glibc="${glibc}" "${container}"
buildah config --label=debian="${debian}" "${container}"

Not sure what did the original buildah do? Just label variable? Was it used by some other where?

martin-g replied:
I don't know where and how these labels are used but they are still set at https://github.com/bioconda/bioconda-containers/pull/55/files#diff-b4a54ac6775437f0dbe546b99c784d76f3defaaa74b1456488e4fa4f78850e1cR75

Yikun replied:
OK, great!

Yikun reviewed (commented):
```
bioconda-utils build --docker --mulled-test --packages py2bit --docker-base-image ghcr.io/yikun/bioconda-utils-build-env-cos7-aarch64 --mulled-conda-image ghcr.io/martin-g/create-env-aarch64 --force
```

I also do a e2e test with https://github.com/martin-g/bioconda-containers/pull/2 (ghcr.io/martin-g/create-env-aarch64), it works! So LGTM except inline comment/question.

martin-g committed:
No need to setup QEMU for create-env for now

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Do not use ${{ steps.build.outputs.tags }} in 'for' loops

Its value is newline separated and breaks the loop syntax

Fix the name of a step - it is steps.build.outputs.image, not
buildah-build anymore

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Add debugging (set -x) to `Check Tags` job

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
Friendly ping!

martin-g commented:
Any feedback on the proposed changes ?

julien-faye commented:
These changes look like exactly what I need!
Let me test them and I'll comment back!

julien-faye commented:
I had to build `bioconda-utils-build-env-cos7-aarch64` myself locally although https://github.com/bioconda/bioconda-utils/pull/866 is merged and I expected to find it at https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7-aarch64

Then I built `create-env-aarch64` by using this PR!

And following the steps explained at https://github.com/bioconda/bioconda-recipes/issues/23454#issuecomment-1521574446 I was able to build several Bioconda packages for Linux aarch64!

Great work, @martin-g !

I hope this PR is merged soon and all aarch64 Docker images are uploaded to Quay.io so that there is no need to use patches and build them locally !

emiliofernandes commented:
+1 for this PR!
Looking forward for it to be merged soon!

johanneskoester commented:
I have never touched this repo so far, so any help with fixing the remaining issues would be greatly appreciated.

martin-g commented:
@johanneskoester Do you have an idea what is needed for https://github.com/bioconda/bioconda-containers/pull/55#issuecomment-1584422904 ?
 https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7-aarch64 gives 404.
The CI job for https://github.com/bioconda/bioconda-utils/pull/866 failed to publish it due to authentication issues.

johanneskoester commented:
> @johanneskoester Do you have an idea what is needed for [#55 (comment)](https://github.com/bioconda/bioconda-containers/pull/55#issuecomment-1584422904) ? https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7-aarch64 gives 404. The CI job for [bioconda/bioconda-utils#866](https://github.com/bioconda/bioconda-utils/pull/866) failed to publish it due to authentication issues.

Thanks for the reminder. Looking into that now. I think the bot did not have the permissions to push, but that should be solved now. Let's see whether the next release (today) succeeds.

johanneskoester commented:
> @johanneskoester Do you have an idea what is needed for [#55 (comment)](https://github.com/bioconda/bioconda-containers/pull/55#issuecomment-1584422904) ? https://quay.io/repository/bioconda/bioconda-utils-build-env-cos7-aarch64 gives 404. The CI job for [bioconda/bioconda-utils#866](https://github.com/bioconda/bioconda-utils/pull/866) failed to publish it due to authentication issues.

Thanks for the reminder. Solved now and 2.5.0 has been properly pushed!

martin-g commented:
Great!
I will update this PR to use it!

martin-g committed:
Update actions/checkout to v4

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
I am going to merge https://github.com/martin-g/bioconda-containers/pull/2 into this PR.

martin-g commented:
The CI jobs fail with:
```
W: The repository 'http://security.debian.org/debian-security stretch/updates Release' does not have a Release file.
W: The repository 'http://deb.debian.org/debian stretch Release' does not have a Release file.
W: The repository 'http://deb.debian.org/debian stretch-updates Release' does not have a Release file.
E: Failed to fetch http://security.debian.org/debian-security/dists/stretch/updates/main/binary-amd64/Packages  404  Not Found [IP: 151.101.2.132 80]
E: Failed to fetch http://deb.debian.org/debian/dists/stretch/main/binary-amd64/Packages  404  Not Found [IP: 146.75.38.132 80]
E: Failed to fetch http://deb.debian.org/debian/dists/stretch-updates/main/binary-amd64/Packages  404  Not Found [IP: 146.75.38.132 80]
E: Some index files failed to download. They have been ignored, or old ones used instead.
```
for [debian:9](https://github.com/martin-g/bioconda-containers/blob/0fc3512b38661fe72575d825b919484d4324f715/images/base-glibc-busybox-bash/Dockerfile#L14-L17).
The comment says that the usage of Debian 9 is intentional (to use old version of Glibc) but the image is broken these days ...

martin-g commented:
Is it OK to use Debian 10 instead ?
Or should I remove the problematic entries in `/etc/apt/sources.list` before executing [apt-get update](https://github.com/bioconda/bioconda-containers/blob/0fc3512b38661fe72575d825b919484d4324f715/images/base-glibc-busybox-bash/Dockerfile#L18) ?

martin-g committed:
Use archive.debian.org for Debian:9

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
With https://github.com/bioconda/bioconda-containers/pull/55/commits/c1e64e218db6e9ae7ed238bdc09a63b671659dbd I updated `sources.list` to use http://archive.debian.org for the `debian:9` Docker image.

martin-g committed:
Do not use `sudo` to install `wget

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g commented:
Now I remember why `create-env` needs to be in a second step:
https://github.com/bioconda/bioconda-containers/blob/0fc3512b38661fe72575d825b919484d4324f715/images/create-env/Dockerfile#L11
here it depends on `quay.io/bioconda/base-glibc-debian-bash` - the ARM64 flavor of this image is created by this PR.

But this PR CI checks cannot pass because https://github.com/bioconda/bioconda-containers/blob/0fc3512b38661fe72575d825b919484d4324f715/.github/workflows/base-glibc-debian-bash.yaml#L96-L98 makes a REST API call to Quay.io that fails with error 403 (authentication) in PRs because Github secrets are not exported to PRs.

martin-g commented:
@johanneskoester Could you please allow the execution of the checks one more time? Thanks!

martin-g committed:
Use apt-get instead of yum. The current BASE is debian

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Run `apt-get update` before trying to install `wget

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

martin-g committed:
Fix a typo in `wget` command option for the output file

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

mbargull commented:
Hi @martin-g,

thank you (and also everyone else above!) for working on this, much appreciated.
I'm currently reviewing this (looking good thus far!) and will help to get these changes published during the next few days.

(Incidentally, Debian 12.2 is also scheduled to be released today, so we might combine updating the base images with that version. I'm doing some sanity checks for that currently.)

I'll get back to you tomorrow or Monday at the latest.

Cheers,
Marcel

mbargull commented:
Quick update:
I'm still checking and debugging some bits and bobs here.

Currently, I'm digging into why `busybox wget` does not work under QEMU (haven't checked on native ARM64 yet).
It's not just our build of BusyBox but also the one from docker-library (both `glibc` and `musl`) and Debian.
Alpine's one work fine (they dyn. link to musl and have a bunch of patches and custom config).
Native `x86_64` work fine, too, of course.
```sh
# podman run --arch=arm64 --rm docker://arm64v8/busybox:1.36.1-glibc busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html
wget: error getting response: Connection reset by peer
# podman run --arch=arm64 --rm docker://arm64v8/busybox:1.36.1-musl busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html
wget: error getting response: Connection reset by peer
# podman run --arch=arm64 --rm docker://arm64v8/debian:12.1-slim sh -c 'apt-get update -qq && DEBIAN_FRONTEND=noninteractive apt-get install -qy busybox > /dev/null && busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html'
debconf: delaying package configuration, since apt-utils is not installed
wget: error getting response: Connection reset by peer
# podman run --arch=arm64 --rm docker://arm64v8/alpine:3.18.4 busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html
  HTTP/1.1 200 OK
  Date: Mon, 09 Oct 2023 20:46:02 GMT
  Server: Apache
  Strict-Transport-Security: max-age=63072000
  Accept-Ranges: bytes
  X-Content-Type-Options: nosniff
  X-Frame-Options: DENY
  X-XSS-Protection: 1
  Connection: close
  Transfer-Encoding: chunked
  Content-Type: text/html
  
# podman run --arch=amd64 --rm docker://amd64/busybox:1.36.1-glibc busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html 
  HTTP/1.1 200 OK
  Date: Mon, 09 Oct 2023 20:46:53 GMT
  Server: Apache
  Strict-Transport-Security: max-age=63072000
  Accept-Ranges: bytes
  X-Content-Type-Options: nosniff
  X-Frame-Options: DENY
  X-XSS-Protection: 1
  Connection: close
  Transfer-Encoding: chunked
  Content-Type: text/html
  
# podman run --arch=amd64 --rm docker://amd64/debian:12.1-slim sh -c 'apt-get update -qq && DEBIAN_FRONTEND=noninteractive apt-get install -qy busybox > /dev/null && busybox wget -q -S --no-check-certificate https://www.busybox.net/news.html'
debconf: delaying package configuration, since apt-utils is not installed
  HTTP/1.1 200 OK
  Date: Mon, 09 Oct 2023 20:47:25 GMT
  Server: Apache
  Strict-Transport-Security: max-age=63072000
  Accept-Ranges: bytes
  X-Content-Type-Options: nosniff
  X-Frame-Options: DENY
  X-XSS-Protection: 1
  Connection: close
  Transfer-Encoding: chunked
  Content-Type: text/html
  
```

Once that one's figured out, I'll cherry-pick your changes into multiple PRs to update the base images and then the other images.

(`docker://debian:12.2` is also not out yet but likely happening in the next days; though we don't have to wait for that.)

martin-g commented:
There are many reports for `busybox` + `wget` + `TLS`: https://github.com/docker-library/busybox/issues?q=Connection+reset+by+peer and https://github.com/mirror/busybox/issues?q=Connection+reset+by+peer

The recommendation is to use `alpine`: https://github.com/docker-library/busybox/issues/64#issuecomment-510194111 and https://github.com/docker-library/busybox/issues/134#issuecomment-1122717673

```
$ docker run --platform=arm64 -it --rm alpine wget -S -c https://www.busybox.net/news.html
```
works fine!

I tried many older versions of `busybox` but all of them fail here.

martin-g committed:
Use ubuntu-22.04 Docker action image for all workflows

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

mbargull commented:
> There are many reports for `busybox` + `wget` + `TLS` \[...\]

Yes, you'd normally not want to use BusyBox's `wget` with TLS etc. without backing it with a well-tested TLS library (it supports using `openssl s_client`).

> The recommendation is to use `alpine` \[...\]

For some use cases that makes sense, yes.
For us not so much. If we were to use Alpine we'd essentially have images with the size of Alpine plus our current image's size (minus the \~800 kB for BusyBox) because we want an image with glibc and Bash (which comprise most of the image's size). It's also not bad to have the two base images somewhat similar.

----
I just wanted to make sure that we don't introduce (avoidable) differences between the `amd64` and `aarch64` images.
I checked on native ARM everything works as expected too -- so all good!
(Turns out the problem is just that BusyBox expects some (not standard-mandated) behavior of `vfork` which QEMU doesn't meet. So, only a problem if someone runs our images via QEMU. We can easily patch that the `wget` applet -- no idea if other (more important) applet are affected too...)

----
I'm currently working on incorporating the version updates and hope we can finish the images tomorrow.

mbargull commented:
After more unfortunate delays, we are nearly there with the container builds.
gh-59 is still building `bioconda/create-env` currently (I need to trim down the tests for that; 10 minutes or so previous run time is north of an hour under the emulator... not very economically...).

In gh-58 I cherry-picked your commits for the base images and those are available on quay.io/bioconda (well, after fixing a little hiccup with gh-60).

There were quite some unexpected hurdles with differences under QEMU and/or aarch64, Buildah versions, bumpy manifest handling, directory layout changes in Debian 12, etc. pp.
The build scripts are, admittedly, a lengthy, under-documented, redundancy-laden mess (I'll try to rectify this in the near-ish future) -- but the image contents seem fine as far as my local inspections/tests and the CI tests show thus far.

I'll wait for one test run of gh-59 to finish and then trim down/skip some tests (at least when run via QEMU), so that we can have those built and pushed today.

We now have x86_64 and aarch64 base images updated to the current Debian/BusyBox versions (also prepared for minimal work needed for future Debian >=13 updates) available at `quay.io/bioconda/base-glibc-busybox-bash:3.0`and `quay.io/bioconda/base-glibc-debian-bash:3.0`.

Thanks very much for keeping to push this -- half a year on this PR, finally coming to fruition!

martin-g commented:
w00t!
#59 is green and (I think) fast enough!

I am going to close this PR since it is done with #58 (and #60).
